### PR TITLE
docs/frontend-system: update extension docs

### PR DESCRIPTION
**Pull request by Rugvip**
Updated extension docs to match the new APIs in 1.30

### CODE REVIEW

File: docs/frontend-system/architecture/20-extensions.md [hunk 1]
```diff
@@ -18,12 +18,12 @@ Each extensions has a number of different properties that define how it behaves
...
 
+When creating an extension do not provide the ID directly. Instead, you indirectly or directly provide the kind, namespace, and name parts that make up the ID. The kind is always provided by the [extension blueprint](./23-extension-blueprints.md), the only exception is if you use [`createExtension`](#creating-an-extensions) directly. Any extension that is provided by a plugin will by default have its namespace set to the plugin ID, so you generally only need to provide an explicit namespace if you want to override an existing extension. The name is also optional, and primarily used to distinguish between multiple extensions of the same kind and namespace. If a plugin doesn't need to distinguish between different extensions of the same kind, the name can be omitted.
```
Comment (freben) [inline docs/frontend-system/architecture/20-extensions.md:23]:
```suggestion
When creating an extension you do not provide the ID directly. Instead, you indirectly or directly provide the kind, namespace, and name parts that make up the ID. The kind is always provided by the [extension blueprint](./23-extension-blueprints.md), the only exception is if you use [`createExtension`](#creating-an-extensions) directly. Any extension that is provided by a plugin will by default have its namespace set to the plugin ID, so you generally only need to provide an explicit namespace if you want to override an existing extension. The name is also optional, and primarily used to distinguish between multiple extensions of the same kind and namespace. If a plugin doesn't need to distinguish between different extensions of the same kind, the name can be omitted.
```

File: docs/frontend-system/architecture/20-extensions.md [hunk 2]
```diff
@@ -64,28 +64,22 @@ const extension = createExtension({
...
-    return {
-      element: <div>Hello World</div>,
-    };
+    return [coreExtensionData.reactElement(<div>Hello World</div>)];
   },
 });
 ```
 
-Note that while the `createExtension` is public API and used in many places, it is not typically what you use when building plugins and features. Instead there are many extension creator functions exported by both the core APIs and plugins that make it easier to create extensions for more specific usages.
+Note that while the `createExtension` is public API and used in many places, it is not typically what you use when building plugins and features. Instead there are many [extension blueprints](./23-extension-blueprints.md) exported by both the core APIs and plugins that make it easier to create extensions for more specific usages.
```
Comment (freben) [inline docs/frontend-system/architecture/20-extensions.md:76]:
```suggestion
Note that while the `createExtension` function is public API and used in many places, it is not typically what you use when building plugins and features. Instead there are many [extension blueprints](./23-extension-blueprints.md) exported by both the core APIs and plugins that make it easier to create extensions for more specific usages.
```

File: docs/frontend-system/architecture/20-extensions.md [hunk 3]
```diff
@@ -101,59 +95,55 @@ The `ExtensionDataRef` can then be used to describe an output property of the ex
...
+Note that you are **not** allowed to repeat the same data reference in the outputs, or return multiple values for the same reference. Multiple outputs for the same reference will conflict with each other and cause an error. If you want to output multiple values of the same TypeScript type you should create separate references for each value. That in turn means that overly generic extension data references a bad idea, for example a generic "string" type. Instead create separate references for each type of data that you want to share.
```
Comment (freben) [inline docs/frontend-system/architecture/20-extensions.md:107]:
```suggestion
Note that you are **not** allowed to repeat the same data reference in the outputs, or return multiple values for the same reference. Multiple outputs for the same reference will conflict with each other and cause an error. If you want to output multiple values of the same TypeScript type you should create separate references for each value. That in turn means that overly generic extension data references are a bad idea, for example a generic "string" type. Instead create separate references for each type of data that you want to share.
```

File: docs/frontend-system/architecture/20-extensions.md [hunk 4]
```diff
@@ -101,59 +95,55 @@ The `ExtensionDataRef` can then be used to describe an output property of the ex
...
   // ...
-  output: {
-    element: coreExtensionData.reactElement.optional(),
-  },
+  output: [coreExtensionData.reactElement.optional()],
   factory() {
-    return {
-      element:
+    return [
+      coreExtensionData.reactElement.optional()(
```
Comment (freben) [inline docs/frontend-system/architecture/20-extensions.md:143]: Really? We shouldn't have to have optional() here too right? Does this example need to be rewritten as a generator or something using like .flat() or so?

File: docs/frontend-system/architecture/20-extensions.md [hunk 5]
```diff
@@ -293,48 +280,68 @@ app:
           title: 'Backstage'
 ```
 
-## Extension Boundary
+## Extension factory as a generator function
+
+In all examples so far we have defined the extension factory as a regular function that returns its output in an array. However, the only requirement is that the factory function returns an iterable of extension data value. This means that you can also define the factory function as a generator function, which allows you to yield values one by one. This is particularly useful if you want to conditionally output values.
```
Comment (freben) [inline docs/frontend-system/architecture/20-extensions.md:285]:
```suggestion
In all examples so far we have defined the extension factory as a regular function that returns its output in an array. However, the only requirement is that the factory function returns any iterable of extension data values. This means that you can also define the factory function as a generator function, which allows you to yield values one by one. This is particularly useful if you want to conditionally output values.
```

File: docs/frontend-system/architecture/20-extensions.md [hunk 6]
```diff
@@ -293,48 +280,68 @@ app:
...
-        ),
-      };
-    },
-  });
-}
+const routableExtension = createExtension({
+  // ...
+  factory({ config, inputs, node }) {
+    const ExtensionComponent = lazy(() =>
+      options
```
Comment (freben) [inline docs/frontend-system/architecture/20-extensions.md:331]: Unclear where this options comes from - this was probably taken from an old extension creator right?

File: docs/frontend-system/architecture/20-extensions.md (hunk 4 above)
Comment (Rugvip) [reply docs/frontend-system/architecture/20-extensions.md]: Ah yep to both of those.

File: docs/frontend-system/architecture/20-extensions.md (hunk 6 above)
Comment (Rugvip) [reply docs/frontend-system/architecture/20-extensions.md]: Yep, needs an update

**Comment by Rugvip** [comment]
Alright, updated

File: docs/frontend-system/architecture/20-extensions.md [hunk 7]
```diff
@@ -293,48 +278,62 @@ app:
...
-      };
-    },
-  });
-}
+const routableExtension = createExtension({
+  // ...
+  factory({ config, inputs, node }) {
+    return [
+      coreExtensionData.path(config.path),
+      coreExtensionData.routeRef(options.routeRef),
```
Comment (freben) [inline docs/frontend-system/architecture/20-extensions.md:330]: more options 😄

**Review by freben — APPROVED** [review]
ay caramba

**Pull request merged**